### PR TITLE
feat: use `kzg-rs` for kzg point evaluation

### PR DESCRIPTION
**Pull request by 0xWOLAND**
Currently, `revm` uses `c-kzg` for KZG point evaluation. It would be useful to have a `[no_std]` endpoint for this, which [`kzg-rs`](https://github.com/0xWOLAND/kzg-rs) can be used for.

### CODE REVIEW

File: crates/precompile/Cargo.toml [hunk 1]
```diff
@@ -46,6 +46,11 @@ bn = { package = "substrate-bn", version = "0.6", default-features = false }
 # KZG point evaluation precompile
 c-kzg = { version = "1.0.2", default-features = false, optional = true }
 
+# Optionally use `kzg-rs` for a pure Rust implementation of KZG point evaluation.
+kzg-rs = { git = "https://github.com/0xWOLAND/kzg-rs", default-features = false, features = [
```
Comment (rakita) [inline crates/precompile/Cargo.toml:50]: needs to be published

Comment (0xWOLAND) [reply crates/precompile/Cargo.toml]: Thanks! I switched to the published version

File: crates/primitives/src/lib.rs [hunk 2]
```diff
@@ -12,7 +12,6 @@ mod constants;
 pub mod db;
 pub mod env;
 
-#[cfg(feature = "c-kzg")]
 pub mod kzg;
```
Comment (rakita) [inline crates/primitives/src/lib.rs:15]:
```suggestion
#[cfg(any(feature = "c-kzg, feature = "kzg-rs"))]
pub mod kzg;
```

Have kzg only if there is any of those features.

File: crates/primitives/src/kzg.rs [hunk 3]
```diff
@@ -1,7 +1,11 @@
 mod env_settings;
 mod trusted_setup_points;
 
+#[cfg(feature = "c-kzg")]
 pub use c_kzg::KzgSettings;
+#[cfg(not(feature = "c-kzg"))]
+pub use kzg_rs::KzgSettings;
```
Comment (rakita) [inline crates/primitives/src/kzg.rs:7]: It would be good to use https://docs.rs/cfg-if/latest/cfg_if/ 
And not having `c-kzg` does not mean that `kzg-rs` is enabled

File: crates/primitives/src/lib.rs [hunk 4]
```diff
@@ -38,10 +38,14 @@ cfg_if::cfg_if! {
...
 pub use kzg::{EnvKzgSettings, KzgSettings};
 pub use precompile::*;
 pub use result::*;
 pub use specification::*;
 pub use state::*;
 pub use utilities::*;
+
+#[cfg(feature = "kzg-rs")]
+// silence lint
+use kzg_rs as _;
```
Comment (rakita) [inline crates/primitives/src/lib.rs:51]: Do you think this needs to be? 
`#[cfg(all(features = "c-kzg",feature = "kzg-rs"))]`

I am assuming here but if both are enabled kzg-rs is not used,

File: crates/precompile/src/lib.rs [hunk 5]
```diff
@@ -22,6 +22,11 @@ pub mod secp256k1;
 pub mod secp256r1;
 pub mod utilities;
 
+// silence lint
+use cfg_if as _;
+#[cfg(feature = "kzg-rs")]
+use kzg_rs as _;
```
Comment (rakita) [inline crates/precompile/src/lib.rs:28]: Similar to: https://github.com/bluealloy/revm/pull/1558/files#r1660049587

**Review by rakita — COMMENTED** [review]
small nits left. Others things lgtm

**Comment by rakita** [comment]
@0xWOLAND can you merge with newest main, PR changes show me diffs of code not related to this PR

File: crates/primitives/Cargo.toml [hunk 6]
```diff
@@ -51,7 +56,7 @@ serde = { version = "1.0", default-features = false, features = [
 hex = { version = "0.4", default-features = false }
 
 [features]
-default = ["std", "c-kzg", "portable"]
+default = ["std", "kzg-rs"]
```
Comment (rakita) [inline crates/primitives/Cargo.toml:59]: default should be `c-kzg`, `kzg-rs` is not audited.

File: crates/primitives/src/kzg/env_settings.rs [hunk 7]
```diff
@@ -2,36 +2,43 @@ use super::{
     trusted_setup_points::{G1_POINTS, G2_POINTS},
     KzgSettings,
 };
+use core::hash::Hash;
```
Comment (rakita) [inline crates/primitives/src/kzg/env_settings.rs:5]: Why is this needed as it was not the previously

**Review by rakita — CHANGES_REQUESTED** [review]
revm/Cargo.toml should have `kzg-rs` feature

**Review by rakita — COMMENTED** [review]
One more thing, feature should be added in revm-precompile

File: crates/precompile/Cargo.toml [hunk 8]
```diff
@@ -98,6 +103,8 @@ secp256r1 = ["dep:p256"]
 
 # Enables the KZG point evaluation precompile.
 c-kzg = ["dep:c-kzg", "revm-primitives/c-kzg"]
+kzg-rs = ["dep:kzg-rs", "revm-primitives/kzg-rs"]
```
Comment (0xWOLAND) [inline crates/precompile/Cargo.toml:106]: > One more thing, feature should be added in revm-precompile
Are you referring to this? I believe that I already included it

Comment (rakita) [reply crates/precompile/Cargo.toml:106]: Not this, check cancun function in iirc lib.rs file. There is recent commit (second or third commit) on main that touches this part. 

There was a condition if ckzg is not enabled to not add precompile.

File: crates/primitives/src/lib.rs [hunk 9]
```diff
@@ -38,10 +38,14 @@ cfg_if::cfg_if! {
...
+#[cfg(any(feature = "c-kzg", feature = "kzg-rs"))]
 pub use kzg::{EnvKzgSettings, KzgSettings};
 pub use precompile::*;
 pub use result::*;
 pub use specification::*;
 pub use state::*;
 pub use utilities::*;
+
+#[cfg(all(feature = "c-kzg", feature = "kzg-rs"))]
+// silence lint
```
Comment (rakita) [inline crates/primitives/src/lib.rs:50]:
```suggestion
// silence kzg-rs lint as c-kzg will be used as default if both are enabled 
```

File: crates/precompile/src/lib.rs [hunk 10]
```diff
@@ -25,6 +25,9 @@ pub mod utilities;
 
 pub use fatal_precompile::fatal_precompile;
 
+#[cfg(feature = "kzg-rs")]
+// silence lint
```
Comment (rakita) [inline crates/precompile/src/lib.rs:29]:
```suggestion
#[cfg(all(feature = "c-kzg", feature = "kzg-rs"))]
// silence kzg-rs lint as c-kzg will be used as default if both are enabled.
use kzg_rs as _;
// silence lint
```

**Review by rakita — APPROVED** [review]
lgtm.

File: crates/primitives/Cargo.toml [hunk 11]
```diff
@@ -105,3 +111,4 @@ rand = ["alloy-primitives/rand"]
 
 # See comments in `revm-precompile`
 c-kzg = ["dep:c-kzg", "dep:once_cell", "dep:derive_more"]
+kzg-rs = ["dep:kzg-rs", "dep:once_cell", "dep:derive_more"]
```
Comment (rakita) [inline crates/primitives/Cargo.toml:114]:
```suggestion
# `kzg-rs` is not audited but useful for `no_std` environment, use it with causing and default to `c-kzg` if possible.
kzg-rs = ["dep:kzg-rs", "dep:once_cell", "dep:derive_more"]
```

File: crates/revm/Cargo.toml [hunk 12]
```diff
@@ -126,6 +126,7 @@ optional_beneficiary_reward = ["revm-interpreter/optional_beneficiary_reward"]
 # See comments in `revm-precompile`
 secp256k1 = ["revm-precompile/secp256k1"]
 c-kzg = ["revm-precompile/c-kzg"]
+kzg-rs = ["revm-precompile/kzg-rs"]
```
Comment (rakita) [inline crates/revm/Cargo.toml:129]:
```suggestion
# `kzg-rs` is not audited but useful for `no_std` environment, use it with causing and default to `c-kzg` if possible.
kzg-rs = ["revm-precompile/kzg-rs"]
```

File: crates/precompile/Cargo.toml (hunk 8 above)
Comment (rakita) [inline crates/precompile/Cargo.toml:106]:
```suggestion
# `kzg-rs` is not audited but useful for `no_std` environment, use it with causing and default to `c-kzg` if possible.
kzg-rs = ["dep:kzg-rs", "revm-primitives/kzg-rs"]
```

**Comment by rakita** [comment]
@0xWOLAND added a few comments for new feature. Will merge PR after CI

File: crates/primitives/src/lib.rs [hunk 13]
```diff
@@ -38,10 +38,14 @@ cfg_if::cfg_if! {
...
+#[cfg(any(feature = "c-kzg", feature = "kzg-rs"))]
 pub use kzg::{EnvKzgSettings, KzgSettings};
 pub use precompile::*;
 pub use result::*;
 pub use specification::*;
 pub use state::*;
 pub use utilities::*;
+
+#[cfg(all(feature = "c-kzg", feature = "kzg-rs"))]
+// silence kzg-rs lint as c-kzg will be used as default if both are enabled
```
Comment (rakita) [inline crates/primitives/src/lib.rs:50]:
```suggestion
// silence kzg-rs lint as c-kzg will be used as default if both are enabled.
```

fmt remove last space.
